### PR TITLE
File io updates

### PR DESCRIPTION
**Pull request by tpfau**
This is a switch to a new SBML IO using libsbml 5.15 instead of 5.13.
It also includes a rework of the readSBML and writeSBML functions. 

**I hereby confirm that I have:**

- [X] Selected `develop` as a target branch (top left drop-down menu)
- [X] Tested my code on my own machine
- [X] Followed the guidelines in the [Contributing Guide](https://github.com/opencobra/cobratoolbox/blob/master/.github/CONTRIBUTING.md)

*(Note: You may replace [ ] with [X] to check the box)*

### CODE REVIEW

File: src/base/io/utilities/SBML/parseSBMLNotesField.m.orig [hunk 1]
```diff
@@ -0,0 +1,120 @@
+<<<<<<< 5933086c821a593ef371fa8475e061683c3d7764:src/base/io/utilities/parseSBMLNotesField.m
```
Comment (laurentheirendt) [inline src/base/io/utilities/SBML/parseSBMLNotesField.m.orig]: @tpfau, may you remove this `.orig` file? Thanks.

File: src/base/io/utilities/SBML/parseSBMLNotesField2012.m.orig [hunk 2]
```diff
@@ -0,0 +1,102 @@
+<<<<<<< 5933086c821a593ef371fa8475e061683c3d7764:src/base/io/utilities/parseSBMLNotesField2012.m
```
Comment (laurentheirendt) [inline src/base/io/utilities/SBML/parseSBMLNotesField2012.m.orig]: @tpfau, may you also remove this `.orig` file?

File: test/verifiedTests/testSBML/testReadSBML.m [hunk 3]
```diff
@@ -52,7 +32,7 @@
 
 
 % load the models
-modelArr = {'Abiotrophia_defectiva_ATCC_49176.xml', 'STM_v1.0.xml', 'iIT341.xml', 'Ec_iAF1260_flux1.xml'};
+modelArr = { 'STM_v1.0.xml', 'iIT341.xml'};
```
Comment (laurentheirendt) [inline test/verifiedTests/testSBML/testReadSBML.m]: I would suggest however to leave `Abiotrophia_defectiva_ATCC_49176.xml'` as a test model.

File: test/verifiedTests/testSBML/testReadSBML.m [hunk 4]
```diff
@@ -78,10 +58,10 @@
 
 
             % define the maximum objective values calculated from pre-converted .mat files
-            modelFBAf_max = [0.149475406282249; 0.477833660760744; 0.692812693473487; 0.736700938865275];
+            modelFBAf_max = [0.477833660760744; 0.692812693473487];
 
             % define the minimum objective values
-            modelFBAf_min = [0.0; 0.0; 0.0; 0.0];
+            modelFBAf_min = [0.0; 0.0];
```
Comment (laurentheirendt) [inline test/verifiedTests/testSBML/testReadSBML.m]: These fields will then have to be updated...

File: test/verifiedTests/testSBML/testReadSBML.m [hunk 5]
```diff
@@ -122,23 +102,12 @@
 
     model.lb = model.lb(:);
     model.ub = model.ub(:);
-
+    model.comps = {'e'};
+    model.compNames = {'ExtraCellular'};
     % add the fields outputted by readSBML
     [model.modelVersion.SBML_level, model.modelVersion.SBML_version, model.modelVersion.fbc_version] = deal(3, 1, 2);
```
Comment (laurentheirendt) [inline test/verifiedTests/testSBML/testReadSBML.m]: What is the reason for removing the `metField` variable?

File: test/verifiedTests/testSBML/testWriteSBML.m [hunk 6]
```diff
@@ -36,15 +45,7 @@
 else
     warning(['The file', fullFileNamePath, ' does not exist and could not be deleted.']);
 end
-
-load('Ec_iJR904.mat')
-model = convertOldStyleModel(model);
```
Comment (laurentheirendt) [inline test/verifiedTests/testSBML/testWriteSBML.m]: What is the reason for removing the test on `convertOldStyleModel`?

File: test/verifiedTests/testSBML/testWriteSBML.m [hunk 7]
```diff
@@ -36,15 +45,7 @@
...
-% use a different deck to convert a COBRA model to an SBML file
-modelSBML = convertCobraToSBML(model);
```
Comment (laurentheirendt) [inline test/verifiedTests/testSBML/testWriteSBML.m]: What is the reason for removing the test on `convertCobraToSBML`?

File: src/base/io/utilities/writeSBML.m [hunk 8]
```diff
@@ -933,21 +800,11 @@
...
-str = strrep(str,',','_COMMA_');
-str = strrep(str,'.','_PERIOD_');
-str = strrep(str,'''','_APOS_');
-str = regexprep(str,'\(e\)$','_e');
-str = strrep(str,'&','&amp;');
-str = strrep(str,'<','&lt;');
-str = strrep(str,'>','&gt;');
-str = strrep(str,'"','&quot;');
-end
+% %% Format For SBML
```
Comment (laurentheirendt) [inline src/base/io/utilities/writeSBML.m]: Do we still need these lines here?

File: src/base/io/utilities/readSBML.m [hunk 9]
```diff
@@ -1,787 +1,494 @@
...
+%extract the fbc_associations fbc_association field or return an empty
+%string
+if isempty(fbc_gprAssoc) || isempty(fbc_gprAssoc.fbc_association)
+    rule = '';
+    return
+end
+try
+    rule = fbc_gprAssoc.fbc_association.fbc_association;
+catch
+    disp('blubb')
```
Comment (laurentheirendt) [inline src/base/io/utilities/readSBML.m]: What does this message mean?

File: src/base/io/utilities/readSBML.m [hunk 10]
```diff
@@ -1,787 +1,494 @@
...
 end
+end
+
+function field = extractfield(startstruct,fieldname)
 
+if ~isfield(startstruct,fieldname)
+    error('Field %s does not exist',fieldname)
+end
+field = {startstruct.(fieldname)};
+%if all(cellfun(@isnumeric , field))
```
Comment (laurentheirendt) [inline src/base/io/utilities/readSBML.m]: Do we still need these lines?

File: src/base/io/utilities/readSBML.m [hunk 11]
```diff
@@ -1,787 +1,494 @@
...
-        if iscell(genes{i})
-            [tmp,geneInd] = ismember(genes{i},allGenes);
+%% Set up the objective. The default is maximisation.
+model.osense = -1;
+if isfield(modelSBML,'fbc_objective')
+    %We only support the first one we find
+    if ~isempty(modelSBML.fbc_objective)
+        osenseStr = modelSBML.fbc_objective(1).fbc_type;
+        if strcmp(osenseStr , 'maximize')
+            model.osense = -1;
```
Comment (laurentheirendt) [inline src/base/io/utilities/readSBML.m]: This condition is not necessary, as `model.osense` has already been initialized to `-1`.

Comment (tpfau) [inline src/base/io/utilities/readSBML.m]: fixed

File: src/base/io/utilities/SBML/parseSBMLNotesField.m.orig (hunk 1 above)
Comment (tpfau) [inline src/base/io/utilities/SBML/parseSBMLNotesField.m.orig]: done

File: src/base/io/utilities/SBML/parseSBMLNotesField2012.m.orig (hunk 2 above)
Comment (tpfau) [inline src/base/io/utilities/SBML/parseSBMLNotesField2012.m.orig]: done

File: src/base/io/utilities/readSBML.m (hunk 9 above)
Comment (tpfau) [inline src/base/io/utilities/readSBML.m]: Leftover from debugging that I forgot to remove.

File: src/base/io/utilities/readSBML.m (hunk 10 above)
Comment (tpfau) [inline src/base/io/utilities/readSBML.m]: no

File: src/base/io/utilities/writeSBML.m (hunk 8 above)
Comment (tpfau) [inline src/base/io/utilities/writeSBML.m]: no

File: test/verifiedTests/testSBML/testReadSBML.m (hunk 3 above)
Comment (tpfau) [inline test/verifiedTests/testSBML/testReadSBML.m]: I put it in again.

File: test/verifiedTests/testSBML/testReadSBML.m (hunk 4 above)
Comment (tpfau) [inline test/verifiedTests/testSBML/testReadSBML.m]: done

File: test/verifiedTests/testSBML/testReadSBML.m (hunk 5 above)
Comment (tpfau) [inline test/verifiedTests/testSBML/testReadSBML.m]: readSBML removes any empty fields. The only fields guaranteed to be present in a model returned by readSBML are (if there are any reactions/mets in the SBML file):
S
lb
ub
c
osense
csense
b
rxns
mets

Everything else could potenitally be removed.

File: test/verifiedTests/testSBML/testWriteSBML.m (hunk 6 above)
Comment (tpfau) [inline test/verifiedTests/testSBML/testWriteSBML.m]: This was not a test for convertOldStyle model but rather something to convert the ecoli model (which is very old). 
But: This test was invalid anyways as it compared an SBML model structure with a COBRA model structure. isSame actually returned false all the time, and it was just that both structs don't have any fields in common that led to numDiff being empty. so essentially, the check for ~any(numDiff) can make sense if one is sure, that both are proper models with potentially different field IDs, but is a bad idea, if that's not sure.

File: test/verifiedTests/testSBML/testWriteSBML.m (hunk 7 above)
Comment (tpfau) [inline test/verifiedTests/testSBML/testWriteSBML.m]: I would alter that function to just call writeSBML (which it kind of does). The inputs of that function were ignored anyways (and it was just kept for backward compatability). Even the old writeSBML was unable to create any SBML Level 2 models and I would strongly discourage giving that option. 
I altered the function accordingly.

**Pull request closed without merging**